Improve: provide options to not pad copied text or include timestamps

SlySven opened the pull request:
This is a solution to the issue raised recently in the Mudlet Discord Help channel which resulted in the creation of the issue that this closes #4956.

The option is presented as a context menu item on any `TConsole` window which is checkable (and defaults to checked to mimic previous behaviour), unchecking it will remove the padding from the first line of a multi-line copy. However if time stamps are shown for the particular console the padding is forced to be on and the context menu option is disabled. Tooltips for both situations are provided.

The option is not saved between sessions because it is individual to each console and as it is so easy to change it is simpler to not try and save it for each and every TConsole instance.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Release post highlight

Provide the option not to pad out the first line of text with spaces when copying multiple lines of text and the selection does not start at the left margin.

SlySven committed:
Enhance: provide option to not pad copied text

This is a solution to the issue raised recently in the Mudlet Discord Help
channel which resulted in the creation of the issue that this closes #4956.

The option is presented as a context menu item on any `TConsole` window
which is checkable (and defaults to checked to mimic previous behaviour),
unchecking it will remove the padding from the first line of a multi-line
copy. However if time stamps are shown for the particular console the
padding is forced to be on and the context menu option is disabled.
Tooltips for both situations are provided.

The option is not saved between sessions because it is individual to each
console and as it is so easy to change it is simpler to not try and save it
for each and every TConsole instance.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
I don't think this is quite the right way, but I do have a solution for this - allow for this to happen via a package: https://github.com/Mudlet/Mudlet/issues/4962

SlySven committed:
Extend: also add control over timestamps being included

Following discussions on Discord the default has been switch to NOT pad
multi-line copies with spaces. The additional option to include timestamps
in the copy defaults to on for each TConsole but is not present in the
context menu until timestamps are showing in the `TConsole` - this is
because only the Main and the Editor Errors for each profile and the
Central Debug console currently display timestamps.

As the previous commit the options are not saved in the profile settings
because it is a per TConsole option and the extra code to save them is, I
feel not worth it when the option is so visible in the context menu.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: keep CodeFactor happy

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_provideoptionToNotPadCopiedText

SlySven commented:
Note that https://github.com/Mudlet/Mudlet/pull/4957/commits/2cc2556a4179ffb888bb13a42284f061b1f9c23b reverts #4959 .

SlySven commented:
:negative_squared_cross_mark: Clang-tidy is unhappy because these two new option states are stored in a pair of `bool`s in the `TConsole` class but is accessed (both read/write) from both `TTextEdit` instances so they share the same setting value. 

The alternative would be to move the flags to the latter and then arrange for any change in one instances value to be replicated in the other - and I think that though this might be better from a pure C++ object encapsulation methodology it is a simpler to do it in the quick and dirty way as I have (and we have done all over the project)...

SlySven commented:
In use - now default (to not pad with spaces) on a console without time stamps:
![image](https://user-images.githubusercontent.com/6163092/111081936-90734d80-84fd-11eb-942d-86ea93a5de1c.png)

default behaviour on a console with time stamps enable:
![image](https://user-images.githubusercontent.com/6163092/111082048-21e2bf80-84fe-11eb-9a61-7ea4ab6548d6.png)

keneanung reviewed (approved):
UI texts are fine by me (didn't check actual C++ code)

SlySven committed:
Merge branch 'development' into Enhance_provideoptionToNotPadCopiedText

vadi2 reviewed (commented):
The "trim timestamps" is a good solution - could we call it "include timestamps" instead though? The positive message is simpler than the negative ones.

As for the space padding - that won't be used often to necessitate it being on the right-click menu.  Let's move it into main display settings instead.

demonnic commented:
> As for the space padding - that won't be used often to necessitate it being on the right-click menu. Let's move it into main display settings instead.

We could easily put out a simple mpackage for this now that strips whitespace from the front of copied text, thanks to https://github.com/Mudlet/Mudlet/pull/5279

vadi2 commented:
True, or even the reverse - keeps it, for those few who'd like this feature

SlySven committed:
Merge branch 'development' into Enhance_provideoptionToNotPadCopiedText

Needed to pull in infrastructure changes since PR initially created.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> The "trim timestamps" is a good solution - could we call it "include timestamps" instead though? The positive message is simpler than the negative ones.

I don't follow you - where does it say "trim timestamps"? The context menu item is called `Timestamps in copy` - which emphasises what operation the item applies to.

> As for the space padding - that won't be used often to necessitate it being on the right-click menu. Let's move it into main display settings instead.

No good - this is a per `TConsole` setting so it belongs on the right click menu for each console - where it can be quickly accessed/seen when copying... the end-user may also want a different setting when they are copying something with time-stamps compared to when not - so it makes sense to keep the two items together when the time-stamp one is shown (only when timestamps ARE enabled for the `TConsole` concerned - which includes the per profile Main and Errors consoles and the Central Debug Console at the moment IIRC).

SlySven committed:
Merge branch 'development' into Enhance_provideoptionToNotPadCopiedText

SlySven committed:
Merge branch 'development' into Enhance_provideoptionToNotPadCopiedText

Kebap commented:
There was a bit discussion in the original issue #4956 as well, but let me summarize my position again:
- We want to let players do easy things easily, not make them overly complicated.
- Copy should work as expected from every other program
- Don't toggle multiple special options in right click menu, then copy suddenly works different
- Instead maybe offer "Copy with padding" and "Copy with/without time-stamps" as a Lua package to the interested few ( I can hardly remember those ever requested )

Not sure if this PR can be adjusted that way or it would require a new PR to be drafted instead,

vadi2 commented:
Agreed 👍

SlySven commented:
Disagreed :-1:

I am not even sure I could code something that implements this an add-in package - it is working on the `TBuffer` instance for any console (not just the main one) and off the top of my head I do not know if we have every bit of what could be needed accessible from the Lua sub-system...

vadi2 commented:
The idea is to make it accessible and then build the package on top of that!

vadi2 commented:
@demonnic @keneanung what do you think of Leris' https://github.com/Mudlet/Mudlet/pull/4957#issuecomment-1021112932?

keneanung commented:
After reading through the PR, I think having checkboxes in context menus that are not saved between sessions is unusual in the least. Especially the not being saved sounds confusing and not user friendly.

Making the context menu of every console pluggable (possibly all at once) and providing the functionality via a new entry for each case sounds much better to me. Then every user can put those options they need together.

demonnic commented:
We can already adjust the right-click context menus using [addMouseEvent](https://wiki.mudlet.org/w/Manual:Lua_Functions#addMouseEvent), I use this for adding them for my highlighter package per category. Implementing a 'copy with padding' should be doable now, and copy with timestamp also doable by using [getTimestamp](https://wiki.mudlet.org/w/Manual:Lua_Functions#getTimestamp)

I think the tools already exist to make the packages now, though they may not have at the beginning of the year.

vadi2 commented:
You're right, tools for this have been added since. I think this could be re-done as a Lua package instead and it would be great for those who'd like this feature.

demonnic commented:
Just so I understand, is the idea

A.) to change the default copy behaviour of Mudlet to be more like other programs and copy without padding or timestamps, and then provide a package which adds 'copy with padding', 'copy with timestamp', and 'copy with padding and timestamp' back to the context menu

or 

B.) The opposite, where we keep current behaviour and provide a package to add 'copy without' options?

I got A vibes from Leris' post, but just wanting to make sure we're all on the same page.

Kebap commented:
A or even multiple packages so players can pick and choose

vadi2 commented:
A as well 

But I haven't heard issues around the behaviour of timestamps, is it wise to change that?

demonnic commented:
The default behaviour around timestamps could be kept, and the package(s) could provide "copy without timestamp" as the option on that side. Timestamps weren't even part of the original issue that I can see.

SlySven commented:
I was trying to provide options to satisfy everyone's copying requirements - and TBH if the rest of you reckon it can be done in Lua rather than by munging around in the C++ core then go right ahead and I'll go and hack in other areas of the codebase...

🤷‍♂️
